### PR TITLE
Implement GL_FRAMEBUFFER_BINDING (and fix a VertexAttrib1fv typo)

### LINKED ISSUE #12852: webgl: Add getParameter(GL_FRAMEBUFFER_BINDING) support
I've been working on adding support for this getter.  Lack of support for it is currently producing a bunch of `INVALID_ENUM` in unrelated webgl tests due to its use in some shared utility code in the tests.  I'm very new to Rust programming, so it's going slowly.

Current code is at:

- https://github.com/anholt/servo/tree/webgl-framebuffer-binding
- https://github.com/anholt/webrender_traits/tree/webgl-framebuffer-binding

(ignore the DO NOT PUSH change in servo, I've been rebasing it out before submission of other patches)

Current questions:

- How does the webrender_traits change get merged safely given that it will produce compile errors in servo until the new enum cases are included?
- What's the right way to wrap my HashMap so that it's mutable at CreateFramebuffer()/DeleteFramebuffer() time, but can still store JS values?
- Is this even the right approach?  My HashMap is storing JS<> values, which I think means that they'll keep the WebGLFramebuffers alive for as long as they're present in the hash.  The intent of WebGL seems to be that they'd be destroyed by the JS GC (once they're also unbound from their binding points in the GL)  This hasn't impacted our other refs because they're only to the current thing in the binding point.

Actually, on the last point: I don't need to be able to map *any* FramebufferId to a framebuffer, I need to be able to map the current bound framebufferid to a framebuffer.  I could track the framebuffer like we do for the current bound program/texture2d/etc!

### PR DESCRIPTION
**Pull request by anholt**
This PR fixes many webgl conformance test failures due to missing support for a getParameter() call in the webgl test utils.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #12852

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12857)

### CODE REVIEW

**Comment by anholt** [comment]
Note: Let me know if including unrelated commits in a PR like this is inappropriate, and I can split it out.

**Comment by emilio** [comment]
@bors-servo: r+

Thanks for doing this @anholt! :)

File: components/script/dom/webglrenderingcontext.rs [hunk 1]
```diff
@@ -1788,7 +1812,7 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
     // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.10
     fn VertexAttrib1fv(&self, _cx: *mut JSContext, indx: u32, data: *mut JSObject) {
         if let Some(data_vec) = array_buffer_view_to_vec_checked::<f32>(data) {
-            if data_vec.len() < 4 {
+            if data_vec.len() < 1 {
```
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: Also, good catch here :)

**Comment by emilio** [comment]
On 08/13/2016 05:37 PM, Eric Anholt wrote:
> Note: Let me know if including unrelated commits in a PR like this is
> inappropriate, and I can split it out.

Oh, also, I don't really mind this, it's fine if it doesn't diverge a
lot from the main code in the PR. Other reviewers might be more picky,
but I think it's not worth the overhead of two different PRs for this :)

Thanks again!

**Comment by emilio** [comment]
```
Tests with unexpected results:
  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/attribs/gl-disabled-vertex-attrib.html:
  │ FAIL [expected PASS] WebGL test #15: at (0, 0) expected: 0,255,0,255 was 255,255,255,255
  │   → assert_true: at (0, 0) expected: 0,255,0,255 was 255,255,255,255 expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1402:20
  │ test@http://web-platform.test:8000/resources/testharness.js:500:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:86:7
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:152:5
  │ checkCanvasRectColor@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1046:9
  │ checkCanvasRect@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1072:3
  │ checkCanvas@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1095:3
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/attribs/gl-disabled-vertex-attrib.html:27:3

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/textures/default-texture.html:
  │ FAIL [expected PASS] WebGL test #0: at (0, 0) expected: 0,0,0,255 was 0,0,128,191
  │   → assert_true: at (0, 0) expected: 0,0,0,255 was 0,0,128,191 expected true got false
  │ 
```

Seems like legit consequences of the new validation an unexpected consequence of the typo fix? I'll double check as fast as I can have a build. Unfortunately the way the wpt and webgl test suite integration can be flaky in this sense (marks not-run tests as passing), but it's in order to minimize pain at the time of updating the test suite.

My intuition is that these tests weren't running without this patch, I'll verify it and if that's the case, I'll r+ once the expectations are updated, thanks!

**Comment by anholt** [comment]
gl-disabled-vertex was me dropping an apparently macos-specific failure on #15.  Can we flag subtests as os-specific failures?

**Comment by emilio** [comment]
@anholt: yes, you can use the `if os == "osx":` syntax that was in the `__DIR__.ini` in individual subsets.

**Comment by anholt** [comment]
And the other one was maybe just #11618 :/

**Comment by emilio** [comment]
Yep! That seems reasonable. r=me with that expectation change

**Comment by anholt** [comment]
Now to go apply this expectation syntax to the "enable webgl testing on linux" series.

**Pull request merged**